File: Hydrogene/Toffee/OS X/Documents/FooDocument.cs
using AppKit;

namespace Documents
{
	[IBObject]
	public class FooDocument : NSDocument
	{
		//
		// NSDocument implementation
		//

		public override this withType(NSString type) error(NSError *error) // workaround for invalid NSError pointer being passed in, sometimes
		{
			data = new FooDocumentData();
		}

		public override bool readFromURL(NSURL! url) ofType(NSString! typeName) error(ref NSError error)
		{
			return loadDocumentFromURL(url);
		}

		public override bool writeToURL(NSURL! url) ofType(NSString! typeName) forSaveOperation(NSSaveOperationType aveOperation) originalContentsURL(NSURL? absoluteOriginalContentsURL) error(ref NSError error)
		{
			return saveDocumentToURL(url);
		}

		public override void makeWindowControllers()
		{
			addWindowController(new FooDocumentWindowController withDocument(self));
		}

		public override static bool autosavesInPlace
		{
			get
			{
				return true; // enables the title bar menu, among other things
			}
		}

		//
		// Custom members
		//

		bool loadDocumentFromURL(NSURL url)
		{
			data = new FooDocumentData withURL(url);
			return data != null;
		}

		bool saveDocumentToURL(NSURL url)
		{
			if (data == null)
			return false;
			data.save(url);
			return true;
		}

		public FooDocumentData data { get; private set; }
	}

	//
	// data stored in the document:
	//

	public class FooDocumentData
	{
		public NSString text { get; set; }

		public this()
		{
			text = "New Foo!";
		}

		public this withURL(NSURL url)
		{
			if (File.Exists(url.path))
			text = File.ReadText(url.path);
			else
			return null;
		}

		public void save(NSURL url)
		{
			File.WriteText(url.path, text);
		}
	}
}

File: Oxygene/Toffee/OS X/Documents/FooDocument.pas
namespace Documents;

interface

uses
  AppKit;

type
  [IBObject]
  FooDocument = public class(NSDocument)
	method loadDocumentFromURL(url: NSURL): Boolean;
	method saveDocumentToURL(url: NSURL): Boolean;
  public
	property data: FooDocumentData read  private write ;

	// NSDocument implementation
	constructor withType(&type: NSString) error(error: ^NSError); override;  // workaround for invalid NSError pointer being passed in, sometimes
	method readFromURL(url: not nullable NSURL) ofType(typeName: not nullable NSString) error(var error: NSError): Boolean; override;
	method writeToURL(url: not nullable NSURL) ofType(typeName: not nullable NSString) forSaveOperation(aveOperation: NSSaveOperationType) originalContentsURL(absoluteOriginalContentsURL: nullable NSURL) error(var error: NSError): Boolean; override;
	method makeWindowControllers; override;
	class property autosavesInPlace: Boolean read true; override;
  end;

  FooDocumentData = public class
  public
	property text: NSString read  write ;
	constructor;
	constructor withURL(url: NSURL);
	method save(url: NSURL);
  end;

implementation

constructor FooDocument withType(&type: NSString) error(error: ^NSError);
begin
  data := new FooDocumentData();
end;

method FooDocument.readFromURL(url: not nullable NSURL) ofType(typeName: not nullable NSString) error(var error: NSError): Boolean;
begin
  result := loadDocumentFromURL(url);
end;

method FooDocument.writeToURL(url: not nullable NSURL) ofType(typeName: not nullable NSString) forSaveOperation(aveOperation: NSSaveOperationType) originalContentsURL(absoluteOriginalContentsURL: nullable NSURL) error(var error: NSError): Boolean;
begin
  result := saveDocumentToURL(url);
end;

method FooDocument.makeWindowControllers;
begin
  addWindowController(new FooDocumentWindowController withDocument(self));
end;

method FooDocument.loadDocumentFromURL(url: NSURL): Boolean;
begin
  data := new FooDocumentData withURL(url);
  exit assigned(data);
end;

method FooDocument.saveDocumentToURL(url: NSURL): Boolean;
begin
  if data = nil then
	exit false;
  data.save(url);
  result := true;
end;

constructor FooDocumentData;
begin
  text := 'New Foo!';
end;

constructor FooDocumentData withURL(url: NSURL);
begin
  if File.Exists(url.path) then
	text := File.ReadText(url.path)
  else
	exit nil;
end;

method FooDocumentData.save(url: NSURL);
begin
  File.WriteText(url.path, text);
end;

end.